### ISSUE TITLE
Composer is not able to resolve dependencies if tightenco/collect is not in __root__

### ISSUE THREAD

**Issue by antonioribeiro** [composer/composer#7129]
I'm using:

```
$ composer --version
Composer version 1.6.3 2018-01-31 16:28:17
```

My `composer.json` is empty

When I run this command:

```
mkdir foo \
 cd foo \
 composer require laravel/framework \
 composer require mathieutu/exporter
```

Composer is not able to resolve it:

```
Writing lock file
Generating autoload files
Using version ^2.0 for mathieutu/exporter
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for mathieutu/exporter ^2.0 -> satisfiable by mathieutu/exporter[2.0.0].
    - Conclusion: remove laravel/framework v5.6.5
    - Conclusion: don't install laravel/framework v5.6.5
    - mathieutu/exporter 2.0.0 requires tightenco/collect ^5.6 -> satisfiable by laravel/framework[v5.6.0, v5.6.1, v5.6.2, v5.6.3], tightenco/collect[v5.6.0, v5.6.4, v5.6.5].
    - Can only install one of: laravel/framework[v5.6.0, v5.6.5].
    - Can only install one of: laravel/framework[v5.6.1, v5.6.5].
    - Can only install one of: laravel/framework[v5.6.2, v5.6.5].
    - Can only install one of: laravel/framework[v5.6.3, v5.6.5].
    - don't install tightenco/collect v5.6.0|don't install laravel/framework v5.6.5
    - don't install tightenco/collect v5.6.4|don't install laravel/framework v5.6.5
    - don't install tightenco/collect v5.6.5|don't install laravel/framework v5.6.5
    - Installation request for laravel/framework (locked at v5.6.5, required as ^5.6) -> satisfiable by laravel/framework[v5.6.5].
```

But if I add tightenco/collect to __root__:

```
mkdir foo \
 cd foo \
 composer require laravel/framework \
 composer require tightenco/collect \
 composer require mathieutu/exporter
```

It works fine.

`mathieutu/exporter` is only requiring "tightenco/collect" and there are no requirements in "tightenco/collect" to any version of laravel/framework. The only relation between `tightenco/collect` and illuminate/framework was a "replace", which is, now, in place only for Collect <5.5.33:

https://github.com/laravel/framework/commit/5caa88acde3d8a82fcbb9bbec8dde45dc49d4b7b#diff-b5d0ee8c97c7abd7e3fa29b9a27d1780L70

Is there anything we can make to fix this?

**Comment by mathieutu** [comment]
@antonioribeiro I think I've understood. As I'm seeing in some issues in this repo (ex: https://github.com/composer/composer/issues/3754) this shape of replace is totally invalid because you can't have several versions of the same package in a project. So `framework` is considered like a old version of `collect`, so you can't install a new version of `collect`...

**Comment by antonioribeiro** [comment]
This is conflicting on a different package vendor, we get the same error installing `laravel/framework` and `spatie/crawler`, but installing `tigthenco/collect` before installing `spatie/crawler` works fine. My opinion is that it should not prevent a package from being installed if the requirement is fulfillable to a single package.

![image](https://user-images.githubusercontent.com/3182864/36909446-bb86ae40-1e1c-11e8-9261-242ba325f460.png)

**Comment by stof** [comment]
> which is, now, in place only for Collect <5.5.33:

The thing is, this still creates a `replace` all the time, but it means that this replacement rule can satisfy any `<5.3.3` requirement of the package, and not higher requirements (while `self.version` means it replaces only a package at the same version).

If the existing `laravel/framework` release don't not **contain** the classes of the collect package, it must not say it replaces it (because that's not true at all). It may be needing to provide a `conflict` rule, but not a `replace`.

Basically, the *only* actual use cases for a `replace` IMO are when exposing subtree of a package as a separate one (symfony components, illuminate packages, etc... which is the reason why the feature was created) and when renaming a package (as a BC layer for people using the old name). In both cases, the right value to use is `self.version` (and any other value should still be a fixed value, as you cannot possibly provide the source code of multiple versions at the same time).

**Comment by mathieutu** [comment]
@stof the problem is that it's used to replace the package until the 5.3.3 version.. :/

**Comment by antonioribeiro** [comment]
> If the existing laravel/framework release don't not contain the classes of the collect package, it must not say it replaces it (because that's not true at all). It may be needing to provide a conflict rule, but not a replace.

To prevent the conflict, the conflicting namespace was changed in 5.5.33 (Collect follows Laravel's version), so now we have two scenarios:

- `laravel/framework >=5.5.33` and `tightenco/collect >=5.5.33`: ` tightenco/collect` **must be** installed because namespaces **are not** conflicting

- `laravel/framework <=5.5.32` and `tightenco/collect <=5.5.32`: `tightenco/collect` **must not** be installed because namespaces **are** conflicting

So, if I understood it well, we should not be using:

``` json
"replace": {
    "tightenco/collect": "<5.5.33"
}
```

But

``` json
"conflict": {
    "tightenco/collect": "<5.5.33"
}
```

This way it would let Composer install `tightenco/collect` and `laravel/framework` only if the selected version is older than `5.5.33`?

**Comment by stof** [comment]
**using** a package does not involve replacing it at all.

your case looks like *old* laravel versions should have the `replace`rule (currently, they have nothing in them so they can be installed in parallel as you want, which will break) and *new* laravel versions should have a conflict rule with older versions.
Fixing things for people running laravel <=5.5.32 is too late (and the `replace` rule in new releases does not change anything for them). But for people running laravel >= 5.3.33, the conflict rule is what is needed (and certainly not a replace as the code is not replaced).

The fact that things work when requiring the package at the root is due to a composer bug when one of the optimizations is partially breaking `replace`

**Comment by antonioribeiro** [comment]
> (currently, they have nothing in them so they can be installed in parallel as you want, which will break)

They (old laravel versions) actually have it set as `self.version`, so it works for versions older than `<5.5.33`. We modified the `replace` to `<5.5.33` when the namespace changed. But, if I understand it right, as `tightenco/collect` is not a subtree of illuminate, it should have been in `conflict` from the very beginning.

Thanks for the clarification, @stof.

**Comment by antonioribeiro** [comment]
Just sent a new PR to fix it, according to what was said in that Composer issue: https://github.com/laravel/framework/pull/23379
